### DISCUSSION TITLE
twinBASIC linguist-language support?

### DISCUSSION THREAD

**Discussion by GCuser99** [DecimalTurn/VBA-on-GitHub#3] [General]
Martin, thanks for doing this research and taking the time to share it with the community!

From your writeup, I read that you aren't using the linguist-language attribute because GH does a good job of correctly associating files with language. But since twinBASIC is not yet a recognized GH language (I don't think it will be long before it does meet the qualification thresholds), what is your advice on using the language attribute for tB files? tB supports .cls and .bas type files, so those won't be GH-distinguishable from VBx. But tB's native .twin holds module/class source text and so if GH user wants to see pretty colorfied code in their repos for those, then maybe consider designating the type as VBA or VB6? The tB native form file is .tbform and is actually JSON.

```
# twinBASIC extensions  - Prevent LF normalization (twin, tbform)
*.[tT][wW][iI][nN]          -text diff linguist-language=VBA
*.[tT][bB][fF][oO][rR][mM]  -text diff linguist-language=JSON
```

I suggest that you place at the top of each .getattribute/.gitignore example files a link back to your GH repo so that they can read your excellent explanations of the issues/solutions - otherwise it will be pilfered from project to project without appreciation! :-). Speaking of that, the (now wholly suboptimal!) .gitattributes file I had BEFORE I read your note to @Cristianbuse, I had pilfered (with clueless abandon) from none-other than VBA-Dictionary, same as the bad-example link that you had sent to Cristain. Ha ha!

Cheers!

**Comment by DecimalTurn** [comment]
Admittedly, when I wrote that GitHub was doing a good enough job at determining if a file was VBA or not, I was thinking mostly of distinguishing VBA and VB6. tB is clearly in a seperate category for the moment, but I think your solution for .twin and .tbfrom files is reasonable while we wait for GH to support tB. You could also do `linguist-language=VB6` for .twin files, but the VB6 grammar in github-linguist still uses the old VB.NET one which is not ideal.

Would you like to make a PR for this? I'd be happy to add this to the template as well.

> I suggest that you place at the top of each .getattribute/.gitignore example files a link back to your GH repo

Yeah, that's probably a good idea, I'll do that soon :). I might even make a release when I feel like it's ready.

> I don't think it will be long before it does meet the qualification thresholds [for tB].

We are currently at 244 files if we exclude fafalone from the search: https://github.com/search?q=path%3A*.twin+%28NOT+IS%3Afork%29+%28NOT+language%3AXML%29+%28NOT+user%3Afafalone%29&type=code
We need at least 2k files for the language to be qualified, so that's already a good amount, but hopefully things will accelerate with the next versions of tB coming up.

**Comment by DecimalTurn** [comment]
@fafalone to follow-up on your [comment](https://github.com/DecimalTurn/VBA-on-GitHub/pull/4#issuecomment-2180571744). 

> I was going by https://github.com/github-linguist/linguist/blob/master/CONTRIBUTING.md and https://github.com/github-linguist/linguist/issues/5756 in particular:
> 
> > at least 2000 files per extension indexed in the last year (the number you see at the top of the search results), unless the extension is expected to only occur once per repo, then 200 files. with a reasonable distribution across unique :user/:repo combinations assessed by manually and randomly clicking through the results.
> 
> You all do need to step up your game though so I'm not so heavily dominating the results 😁 
> 
> Actually speaking of that; nothing comes up on a search for .twinproj extension (lots for vbp); I don't know what's up there, do you have to have a recognized extension for that to count? Or not be binary?

The fact that the .twin file does not "only occur once per repo" makes me think the files count is the rule that will matter in this case.

However, you are more than right that the community needs to step up their game 😅 because no matter the metric we use at the moment, it's not there yet. 

Regarding .twinproj files, I had the same issue ([no results](https://github.com/search?q=path%3A*.twinproj%20&type=code)), but I have no idea why.

**Comment by fafalone** [comment]
370 if you don't exclude me, so might be I'm not dominant enough to get excluded...

twinproj occurs once per repo, so I thought of using that. That would give 'number of projects'

Hopefully things will take off with the big v1.0 release.

**Reply by DecimalTurn** [reply]
Ok, I see your reasoning now. The problem with the .twinproj file is that it's binary, so it can't be associated to a language in Linguist. Only code files can be associated to a language. 

The goal of Linguist is to provide syntax highlighting and language statistics, so it doesn't make sense in the context of a binary file or a file that doesn't contain code in the language it's associated to (see for reference this comment: https://github.com/github-linguist/linguist/issues/5824#issuecomment-1296234904)
